cloud/amazon/ec2_asg.py

```diff
@@ -226,428 +218,557 @@
-
-    if not as_groups:
+        for k, v in tag.iteritems():
+            if k != 'propagate_at_launch':
+                asg_tags.append(dict(Key=k,
+                                     Value=v,
+                                     PropagateAtLaunch=bool(tag.get('propagate_at_launch', True)),
+                                     ResourceType='auto-scaling-group',
+                                     ResourceId=group_name))
+    if not as_groups.get('AutoScalingGroups'):
         if not vpc_zone_identifier and not availability_zones:
-            availability_zones = module.params['availability_zones'] = [zone.name for zone in ec2_connection.get_all_zones()]
+            availability_zones = module.params['availability_zones'] = [zone.name for zone in
+                                                                        ec2_connection.get_all_zones()]
         enforce_required_arguments(module)
-        launch_configs = connection.get_all_launch_configurations(names=[launch_config_name])
-        if len(launch_configs) == 0:
-            module.fail_json(msg="No launch config found with name %s" % launch_config_name)
-        ag = AutoScalingGroup(
-                 group_name=group_name,
-                 load_balancers=load_balancers,
-                 availability_zones=availability_zones,
-                 launch_config=launch_configs[0],
-                 min_size=min_size,
-                 max_size=max_size,
-                 placement_group=placement_group,
-                 desired_capacity=desired_capacity,
-                 vpc_zone_identifier=vpc_zone_identifier,
-                 connection=connection,
-                 tags=asg_tags,
-                 health_check_period=health_check_period,
-                 health_check_type=health_check_type,
-                 default_cooldown=default_cooldown,
-                 termination_policies=termination_policies)
+        launch_configs = connection.describe_launch_configurations(LaunchConfigurationNames=[launch_config_name])
+        ag = dict(
+            AutoScalingGroupName=group_name,
+            AvailabilityZones=availability_zones,
+            LaunchConfigurationName=launch_configs['LaunchConfigurations'][0]['LaunchConfigurationName'],
+            MinSize=min_size,
+            MaxSize=max_size,
+            DesiredCapacity=desired_capacity,
+            VPCZoneIdentifier=vpc_zone_identifier,
+            Tags=asg_tags,
+            HealthCheckGracePeriod=health_check_period,
+            HealthCheckType=health_check_type,
+            DefaultCooldown=default_cooldown,
+            TerminationPolicies=termination_policies)
+        if load_balancers:
+            ag['LoadBalancerNames'] = load_balancers
+        if target_group_arns:
+            ag['TargetGroupARNs'] = target_group_arns
 
         try:
-            connection.create_auto_scaling_group(ag)
-            suspend_processes(ag, module)
+            connection.create_auto_scaling_group(**ag)
```

Comment:
Instead of just the try/except, can you use the new AWSRetry library that's in ec2's module_utils for external calls like this? It handles backoffs/retries automatically.